README.md
<p align="center">
  <img src="https://raw.githubusercontent.com/furcan/BetterBoard/main/docs/github-cover.png" width="640" height="auto" alt="BetterBoard">
</p>

[npm-version-badge]: https://img.shields.io/npm/v/BetterBoard.svg
[npm-version-url]: https://www.npmjs.com/package/BetterBoard
[synk-badge]: https://snyk.io/test/github/furcan/BetterBoard/badge.svg?targetFile=package.json
[synk-url]: https://snyk.io/test/github/furcan/BetterBoard?targetFile=package.json
[typescript-badge]: https://badgen.net/badge/icon/Typed?icon=typescript&label&labelColor=3178c6&color=555
[typescript-url]: https://github.com/furcan/BetterBoard/blob/main/index.d.ts
[lic-badge]: https://img.shields.io/github/license/furcan/BetterBoard.svg
[lic-url]: https://github.com/furcan/BetterBoard/blob/main/LICENSE

[![NPM Version][npm-version-badge]][npm-version-url]
[![Known Vulnerabilities][synk-badge]][synk-url]
[![TypeScript][typescript-badge]][typescript-url]
[![License][lic-badge]][lic-url]

# BetterBoard - The ULTIMATE virtual keyboard solution
A pure JavaScript library for using virtual keyboards.

---------

## Current Version
1.0.0-beta.01 [*](https://github.com/furcan/BetterBoard/blob/main/CHANGELOG.md)

---------

## Documentation and Demo
COMING SOON!

---------

## Browser Compatibility
`Chrome` || `Firefox` || `Safari` || `Opera` || `Edge` || `IE 11`

---------
## (A) Install & Import

### Install
COMING SOON... None of these work btw
#### [yarn](https://yarnpkg.com/package/BetterBoard)
```js
yarn add BetterBoard
```

#### [npm](https://www.npmjs.com/package/BetterBoard)
```js
npm i BetterBoard
```
### Import

```jsx
import BetterBoard from 'BetterBoard';
```

---------

## (B) Adding to an HTML Document
### CSS and JS
Actually works!:
```html
<script src="https://cdn.jsdelivr.net/gh/Sid220/BetterBoard@master/src/betterboard.js"></script>
```

### Or only JS (All in One - Internal CSS)

```html
<script src="dist/BetterBoard-aio-2.1.0.min.js"></script>
```

---------


## Keyboard Types and Themes
3 types of keyboards can be used: `all`, `keyboard`, and `numpad`.

5 types of themes can be used. `light`, `dark`, `flat`, `material`, and `oldschool`.

---------

## Run / Initialize
BetterBoard Virtual Keyboard can be used with the `input` or `textarea` elements.

BetterBoard must be initialized with the required options. The other ones are optional. Keyboard Type (the default value is "all") `data-BetterBoard-type`, Keyboard Placement (the default value is "bottom") `data-BetterBoard-placement`, and Special Characters `data-BetterBoard-specialcharacters` settings are each element-based (data attributes).

All options and examples of data attribute usages are as below. Also, a custom class name can be defined as globally for all input and/or textarea elements to run BetterBoard.

---------

### HTML => (data-* options)
```html
<!-- An example of a textarea element: The keyboard type is "all", the placement is "top", and the availability of the special characters is "true". -->
<textarea class="js-virtual-keyboard" data-BetterBoard-type="all" data-BetterBoard-placement="top" data-BetterBoard-specialcharacters="true" placeholder="Your Address"></textarea>

<!-- An example of an input element: The keyboard type is "keyboard", the placement is "bottom", and the availability of the special characters is "false". -->
<input class="js-virtual-keyboard" data-BetterBoard-type="keyboard" data-BetterBoard-placement="bottom" data-BetterBoard-specialcharacters="false" placeholder="Your Name" />

<!-- An example of an input element: The keyboard type is "numpad", and the placement is "bottom". (Special characters are not allowed for "numpad".) -->
<input class="js-virtual-keyboard" data-BetterBoard-type="numpad" data-BetterBoard-placement="bottom" placeholder="Your Number" />
```

---------

### JS => (Run with Init)

```js
// Select the input or the textarea element(s) to run the BetterBoard

BetterBoard.run('.js-virtual-keyboard', {
   // ...init options
});
```

### OR

### JS => (Step1: Initialize)

```js
// Initialize BetterBoard (default/all options)

BetterBoard.init({

  /*!
  * Required
  * An Array of Objects has to be defined for the custom keys. Hint: Each object creates a row element (HTML) on the keyboard.
  * e.g. [{"key":"value"}, {"key":"value"}] => [{"0":"A","1":"B","2":"C"}, {"0":"D","1":"E","2":"F"}]
  */
  keysArrayOfObjects: null,

  /*!
  * Required only if "keysArrayOfObjects" is "null".
  * The path of the "BetterBoard-keys-${langugage}.json" file must be set to the "keysJsonUrl" option. (XMLHttpRequest to get the keys from JSON file.)
  * e.g. '/Content/Plugins/BetterBoard/dist/BetterBoard-keys-english.json'
  */
  keysJsonUrl: null,

  /*
  * Optional: An Array of Strings can be set to override the built-in special characters.
  * e.g. ["#", "€", "%", "+", "-", "*"]
  */
  keysSpecialCharsArrayOfStrings: null,

  /*
  * Optional: An Array of Numbers can be set to override the built-in numpad keys. (From 0 to 9, in any order.)
  * e.g. [1, 2, 3, 4, 5, 6, 7, 8, 9, 0]
  */
  keysNumpadArrayOfNumbers: null,

  // Optional: (Other Options)

  // Language Code (ISO 639-1) for custom keys (for language support) => e.g. "de" || "en" || "fr" || "hu" || "tr" etc...
  language: 'en',

  // The theme of keyboard => "light" || "dark" || "flat" || "material" || "oldschool"
  theme: 'light',

  // Uppercase or lowercase to start. Uppercased when "true"
  capsLockActive: true,

  /*
  * Allow or prevent real/physical keyboard usage. Prevented when "false"
  * In addition, the "allowMobileKeyboard" option must be "true" as well, if the real/physical keyboard has wanted to be used.
  */
  allowRealKeyboard: false,

  // Allow or prevent mobile keyboard usage. Prevented when "false"
  allowMobileKeyboard: false,

  // CSS animations for opening or closing the keyboard
  cssAnimations: true,

  // CSS animations duration as millisecond
  cssAnimationsDuration: 360,

  // CSS animations style for opening or closing the keyboard => "slide" || "fade"
  cssAnimationsStyle: 'slide',

  // Enable or Disable Spacebar functionality on the keyboard. The Spacebar will be passive when "false"
  keysAllowSpacebar: true,

  // Text of the space key (Spacebar). Without text => " "
  keysSpacebarText: 'Space',

  // Font family of the keys
  keysFontFamily: 'sans-serif',

  // Font size of the keys
  keysFontSize: '22px',

  // Font weight of the keys
  keysFontWeight: 'normal',

  // Size of the icon keys
  keysIconSize: '25px',

  // Scrolls the document to the top or bottom(by the placement option) of the input/textarea element. Prevented when "false"
  autoScroll: true,
});
```

### JS => (Step2: Run)

```js
// Select the input or the textarea element(s) to run the BetterBoard

BetterBoard.run('.js-virtual-keyboard');
```

---------

## Language (JSON)
The `keysJsonUrl` option has to be set if custom keys are not defined with the `keysArrayOfObjects` option. JSON format has to be: `[{"key":"value", "key":"value"}, ...]`. Each object in that array creates a row element (HTML) on the keyboard. The "key" in the objects is used as an "index" for each Keyboard Keys. The "value" is each key's value and inner text.

Additionally, BetterBoard includes 9 different language packages: `Arabic`, `English`, `French`, `German`, `Hungarian`, `Persian`, `Russian`, `Spanish`, and `Turkish`.

An example of a JSON file (for custom keys) in English.
```json
[
   {
      "0": "Q",
      "1": "W",
      "2": "E",
      "3": "R",
      "4": "T",
      "5": "Y",
      "6": "U",
      "7": "I",
      "8": "O",
      "9": "P"
   },
   {
      "0": "A",
      "1": "S",
      "2": "D",
      "3": "F",
      "4": "G",
      "5": "H",
      "6": "J",
      "7": "K",
      "8": "L"
   },
   {
      "0": "Z",
      "1": "X",
      "2": "C",
      "3": "V",
      "4": "B",
      "5": "N",
      "6": "M"
   }
]

```

---------
---------
---------

## Copyright
Copyright © 2022 BetterBoard - Virtual Keyboard

## License
MIT license - https://opensource.org/licenses/MIT


CHANGELOG.md
@1.0.0-beta.01
* BetterBoard. Coming Soon.

-----
@1.0.0-beta.02 & 3
* BetterBoard. Stupid mistake and I needed to publish to npm again...

-----
@1.0.0-beta.06 & 7 & 8 & 9 
* Added closeKeyboard() function! Yay!
* Added isKeyboardOpen() function! Yay! Returns true if keyboard is open.

-----


index.d.test.js
import BetterBoard from 'index.d';
BetterBoard.init({
    keysArrayOfObjects: [
        {
            "0": "S",
            "1": "O",
            "2": "L"
        },
        {
            "0": "L",
            "1": "U",
            "2": "C",
            "3": "E",
            "4": "T"
        },
        {
            "0": "O",
            "1": "M",
            "2": "N",
            "3": "I",
            "4": "B",
            "5": "U",
            "6": "S"
        },
    ],
    keysJsonUrl: '/Content/Plugins/BetterBoard/dist/BetterBoard-keys-english.json',
    keysSpecialCharsArrayOfStrings: ["#", "€", "%", "+", "-", "*"],
    keysNumpadArrayOfNumbers: [1, 2, 3, 4, 5, 6, 7, 8, 9, 0],
    language: 'en',
    theme: 'light',
    capsLockActive: true,
    allowRealKeyboard: false,
    allowMobileKeyboard: false,
    cssAnimations: true,
    cssAnimationsDuration: 360,
    cssAnimationsStyle: 'slide',
    keysAllowSpacebar: true,
    keysSpacebarText: 'Space',
    keysFontFamily: 'sans-serif',
    keysFontSize: '22px',
    keysFontWeight: 'normal',
    keysIconSize: '25px',
    autoScroll: true,
    keyboardType: 'numpad'
});
BetterBoard.run('.js-BetterBoard', {
    keysArrayOfObjects: [
        {
            "0": "S",
            "1": "O",
            "2": "L"
        },
        {
            "0": "L",
            "1": "U",
            "2": "C",
            "3": "E",
            "4": "T"
        },
        {
            "0": "O",
            "1": "M",
            "2": "N",
            "3": "I",
            "4": "B",
            "5": "U",
            "6": "S"
        },
    ],
    allowRealKeyboard: false,
    allowMobileKeyboard: false,
    language: 'en',
    theme: 'dark',
    keyboardType: 'all'
});
var input = window.document.querySelector('.js-BetterBoard');
BetterBoard.run(input, {
    keysArrayOfObjects: [
        {
            "0": "F",
            "1": "U",
            "2": "R",
            "3": "C",
            "4": "A",
            "5": "N"
        },
    ],
    allowRealKeyboard: false,
    allowMobileKeyboard: true,
    language: 'en',
    theme: 'dark',
    keyboardType: 'keyboard'
});


index.d.test.ts
import BetterBoard from 'index.d';

BetterBoard.init({
  keysArrayOfObjects: [
    {
      "0": "S",
      "1": "O",
      "2": "L",
    },
    {
      "0": "L",
      "1": "U",
      "2": "C",
      "3": "E",
      "4": "T",
    },
    {
      "0": "O",
      "1": "M",
      "2": "N",
      "3": "I",
      "4": "B",
      "5": "U",
      "6": "S",
    },
  ],
  keysJsonUrl: '/Content/Plugins/BetterBoard/dist/BetterBoard-keys-english.json',
  keysSpecialCharsArrayOfStrings: ["#", "€", "%", "+", "-", "*"],
  keysNumpadArrayOfNumbers: [1, 2, 3, 4, 5, 6, 7, 8, 9, 0],
  language: 'en',
  theme: 'light',
  capsLockActive: true,
  allowRealKeyboard: false,
  allowMobileKeyboard: false,
  cssAnimations: true,
  cssAnimationsDuration: 360,
  cssAnimationsStyle: 'slide',
  keysAllowSpacebar: true,
  keysSpacebarText: 'Space',
  keysFontFamily: 'sans-serif',
  keysFontSize: '22px',
  keysFontWeight: 'normal',
  keysIconSize: '25px',
  autoScroll: true,
  keyboardType: 'numpad',
});

BetterBoard.run('.js-BetterBoard', {
  keysArrayOfObjects: [
    {
      "0": "S",
      "1": "O",
      "2": "L",
    },
    {
      "0": "L",
      "1": "U",
      "2": "C",
      "3": "E",
      "4": "T",
    },
    {
      "0": "O",
      "1": "M",
      "2": "N",
      "3": "I",
      "4": "B",
      "5": "U",
      "6": "S",
    },
  ],
  allowRealKeyboard: false,
  allowMobileKeyboard: false,
  language: 'en',
  theme: 'dark',
  keyboardType: 'all',
});

const input = window.document.querySelector('.js-BetterBoard') as HTMLInputElement;
BetterBoard.run(input, {
  keysArrayOfObjects: [
    {
      "0": "F",
      "1": "U",
      "2": "R",
      "3": "C",
      "4": "A",
      "5": "N",
    },
  ],
  allowRealKeyboard: false,
  allowMobileKeyboard: true,
  language: 'en',
  theme: 'dark',
  keyboardType: 'keyboard',
});


.dev/dev-allinone.js
/*!
* All In One Generator
*
* Description: Creates "BetterBoard-aio.js" file automatically from "BetterBoard.js" and "BetterBoard.css" files.
* Version: 2.1.0
* Author: Furkan MT ('https://github.com/furcan')
* Copyright 2022 All In One Generator, MIT Licence ('https://opensource.org/licenses/MIT')
*/

// Dev Dependencies
const { existsSync, readFileSync, writeFileSync } = require('fs');
const { join } = require('path');
const CleanCSS = require('clean-css');
const Constants = require('./dev-constants');

// Constants
const thisFilePath = '.dev/dev-allinone.js';

// Get File content as text: begin
const getFileContentAsTextByType = (filePath, fileType, clean) => {
  // if file is exist
  if (existsSync(filePath)) {
    // file as text
    const fileAsText = readFileSync(filePath, 'utf-8');
    // if file contains text
    if (typeof fileAsText === 'string' && fileAsText.length > 0) {
      let code = fileAsText;
      if (fileType === 'style' && clean) {
        const styleClean = new CleanCSS(Constants.cleanCSSOptions).minify(fileAsText);
        code = (styleClean || {}).styles;
      }
      return code || false;
    } else {
      Constants.terminalError(`"${filePath}" file is empty and/or something went wrong.`, `${thisFilePath} => Line: 34`);
      return false;
    }
  }
  // else throw error
  else {
    Constants.terminalError(`"${filePath}" directory does not exist in the root directory.`, `${thisFilePath} => Line: 40`);
    return false;
  }
};
// Get File content as text: end

// Create "BetterBoard-aio.js" file from "BetterBoard.js" and "BetterBoard.css" files: begin
const createAIOfileFromJsAndCss = () => {
  // BetterBoard style as minified
  const stylePath = join(Constants.dirInputDev, Constants.fileStyle);
  const styleAsMinified = getFileContentAsTextByType(stylePath, 'style', true);

  // BetterBoard script
  const scriptPath = join(Constants.dirInputDev, Constants.fileScript);
  const scriptAsNormal = getFileContentAsTextByType(scriptPath, 'script', false);

  // if style and script are exist
  if (styleAsMinified && scriptAsNormal) {
    // if output directory is exist
    if (existsSync(Constants.dirOutputDev)) {
      const commentVersion = `* Author: The Fake Slim Shady ('https://github.com/Sid220')`
      const commentDescAndVersion = `* Description: This file contains the BetterBoard CSS codes as internal to use the BetterBoard as one file. This file has been created automatically from using the "BetterBoard.js", and "BetterBoard.css" files.\n${commentVersion}`;
      const internalCSS = `var internalCSS = '';`;
      // add a description comment before the version && replace internal css codes => if they exist
      if (scriptAsNormal.indexOf(commentVersion) > -1 && scriptAsNormal.indexOf(internalCSS) > -1) {
        // add the description above the version as comment
        let scriptAsNormalAIO = scriptAsNormal.replace(commentVersion, commentDescAndVersion);
        // replace internal css
        scriptAsNormalAIO = scriptAsNormalAIO.replace(internalCSS, `var internalCSS = '${styleAsMinified}';`);
        // create "BetterBoard-aio.js" file
        writeFileSync(join(Constants.dirOutputDev, Constants.fileScriptAIO), scriptAsNormalAIO);
      } else {
        Constants.terminalError(`"${scriptAsNormal.indexOf(commentVersion) === -1 ? commentVersion : internalCSS}" does not exist in the "${Constants.fileScript}" file.`, `${thisFilePath} => Line: 72`);
        return false;
      }
    }
    // else throw error
    else {
      Constants.terminalError(`"${Constants.dirOutputDev}" directory does not exist in the root directory.`, `${thisFilePath} => Line: 78`);
      return false;
    }
  } else {
    Constants.terminalError(`Something went wrong on ${!styleAsMinified ? '"styleAsMinified"' : '"scriptAsNormal"'}.`, `${thisFilePath} => Line: 82`);
    return false;
  }
};
createAIOfileFromJsAndCss();
// Create "BetterBoard-aio.js" file from "BetterBoard.js" and "BetterBoard.css" files: end


.dev/dev-minifier.js
/*!
* Minifier
*
* Description: Minify the BetterBoard scripts, and clean the BetterBoard styles to the distribution. (Used "Babel Minify", and "Clean CSS")
* Version: 2.1.0
* Author: Furkan MT ('https://github.com/furcan')
* Copyright 2022 Minifier, MIT Licence ('https://opensource.org/licenses/MIT')
*/

// Dev Dependencies
const { existsSync, readFileSync, writeFileSync } = require('fs');
const { join } = require('path');
const Minify = require('babel-minify');
const CleanCSS = require('clean-css');
const Constants = require('./dev-constants');

// Constants
const thisFilePath = '.dev/dev-minifier.js';

// Minified Code and Source Type: begin
const minifiedCodeBySourceType = (text, type) => {
  // if script
  if (type === 'script') {
    const script = Minify(text, Constants.minifyOptions, Constants.minifyOverrides);
    if (typeof script === 'object' && typeof script.code === 'string' && typeof script.sourceType === 'string') {
      return {
        code: script.code,
        type: script.sourceType,
      };
    } else {
      return false;
    }
  }
  // else if style
  else if (type === 'style') {
    const style = new CleanCSS(Constants.cleanCSSOptions).minify(text);
    if (typeof style === 'object' && typeof style.styles === 'string') {
      return {
        code: style.styles,
        type: type,
      };
    } else {
      return false;
    }
  }
  // else
  return false;
};
// Minified Code and Source Type: end

// Write The File Into The Output Directory: begin
const writeFileToTheOutDir = (minContent, fileName, filePath) => {
  if (existsSync(Constants.dirOutputDist)) {
    // create "BetterBoard.min.*" file by minified content
    if (typeof minContent === 'object' && (typeof minContent.type === 'string' && (typeof minContent.code === 'string' && minContent.code.length > 0))) {
      // file extention
      let ext = null;
      if (minContent.type === 'script') {
        ext = 'js';
      } else if (minContent.type === 'style') {
        ext = 'css';
      }

      // if ext exist create a file
      if (typeof ext === 'string') {
        // comment line
        const comment = `/* ${Constants.title} ${Constants.url} - Version: ${Constants.version} - Author: ${Constants.author} - Copyright ${Constants.year} ${Constants.title}, ${Constants.license} */\n\n`;
        // minified code with comment line
        const code = comment + minContent.code;
        // minified file name
        const minFileName = `${fileName}-${Constants.version}.min.${ext}`;
        // create a minified file into the output directory
        writeFileSync(join(Constants.dirOutputDist, minFileName), code);
      }

    } else {
      Constants.terminalError(`The "${filePath}" file is empty and/or something went wrong.`, `${thisFilePath} => Line: 77`);
      return false;
    }
  } else {
    Constants.terminalError(`The "${Constants.dirOutputDist}" directory does not exist in the root directory.`, `${thisFilePath} => Line: 81`);
    return false;
  }
};
// Write The File Into The Output Directory: end

// Create The File from by The Input Directory: begin
const createFileFromInputDir = (filePath, fileName, filePrefix, fileType) => {
  if (existsSync(filePath)) { // if file exist
    // file text
    const fileText = readFileSync(filePath, 'utf-8');
    // if file text exist
    if (typeof fileText === 'string' && fileText.length > 0) {
      // minified content by file
      const minifiedContent = minifiedCodeBySourceType(fileText, fileType);
      // create a file by minified content
      writeFileToTheOutDir(minifiedContent, filePrefix, fileName);
    }
    // else throw error
    else {
      Constants.terminalError(`The "${filePath}" file is empty and/or something went wrong.`, `${thisFilePath} => Line: 101`);
      return false;
    }
  } else {
    Constants.terminalError(`The "${fileName}" file does not exist in the "${Constants.dirInputDev}" directory.`, `${thisFilePath} => Line: 105`);
    return false;
  }
};
// Create The File from by The Input Directory: end

// Minify: begin
if (existsSync(Constants.dirInputDev) && existsSync(Constants.dirOutputDev)) { // if the input directory exist
  // Script: begin
  const scriptPath = join(Constants.dirInputDev, Constants.fileScript);
  createFileFromInputDir(scriptPath, Constants.fileScript, Constants.prefix, 'script');
  // Script: end

  // Style: begin
  const stylePath = join(Constants.dirInputDev, Constants.fileStyle);
  createFileFromInputDir(stylePath, Constants.fileStyle, Constants.prefix, 'style');
  // Style: end

  // Script (All In One): begin
  const scriptAIOPath = join(Constants.dirOutputDev, Constants.fileScriptAIO);
  createFileFromInputDir(scriptAIOPath, Constants.fileScriptAIO, `${Constants.prefix}-aio`, 'script');
  // Script (All In One): end
} else {
  Constants.terminalError(`The "${!existsSync(Constants.dirInputDev) ? Constants.dirInputDev : Constants.dirOutputDev}" directory does not exist in the root directory.`, `${thisFilePath} => Line: 128`);
  return false;
}
// Minify: end


.dev/dev-constants.js
/*!
* Constants
* Description: Constants for the development.
* Version: 2.1.0
* Author: Furkan MT ('https://github.com/furcan')
* Copyright 2022 Constants, MIT Licence ('https://opensource.org/licenses/MIT')
*/

// Dev Dependencies
const chalk = require('chalk');
const package = require('../package.json');
var fs = require('fs');

// Constants: begin
// - CleanCSS Options: begin
const cleanCSSOptions = {
  level: {
    1: {
      optimizeBackground: false, // controls `background` property optimizations; defaults to `true`
      optimizeBorderRadius: false, // controls `border-radius` property optimizations; defaults to `true`
      optimizeFilter: false, // controls `filter` property optimizations; defaults to `true`
      optimizeFont: false, // controls `font` property optimizations; defaults to `true`
      optimizeFontWeight: false, // controls `font-weight` property optimizations; defaults to `true`
      optimizeOutline: false, // controls `outline` property optimizations; defaults to `true`
      specialComments: false, // remove all comments
      removeQuotes: false, // controls removing quotes when unnecessary; defaults to `true`
      semicolonAfterLastProperty: true, // controls removing trailing semicolons in rule; defaults to `false` - means remove
    },
  },
};
// - CleanCSS Options: end

// - Babel Minify Options: begin
const minifyOptions = {
  builtIns: false, // transform-minify-booleans
};

const minifyOverrides = {
  comments: false, // remove all comments
};
// - Babel Minify Options: end

// - Terminal Error Message: begin
const terminalError = (message, fileOrPath) => {
  const colorRed = '#ff5549';
  const colorBlue = '#26c0d3';
  if (typeof fileOrPath !== 'string') { fileOrPath = '???'; }
  let info = chalk.hex(colorBlue)('\nPlease look at the "' + fileOrPath + '" for more information.\n\n');
  if (typeof message !== 'string') {
    message = 'An error has occurred on: "' + chalk.hex(colorBlue)(fileOrPath) + '"';
    info = '';
  }
  return console.error(chalk.hex(colorRed).bold('Development Error: ') + chalk.hex(colorRed)(message) + info);
};
// - Terminal Error Message: end

// - Exports: begin
module.exports = {
  prefix: 'BetterBoard',
  dirInputDev: 'src',
  dirOutputDev: 'build',
  dirOutputDist: 'dist',
  fileScript: 'betterboard.js',
  fileStyle: 'betterboard.css',
  fileScriptAIO: 'betterboard-aio.js',
  version: (JSON.stringify((package || {}).version) || '').replace(/"/gm, ''),
  author: (JSON.stringify((package || {}).author) || '').replace(/"/gm, ''),
  title: 'BetterBoard - Virtual Keyboard',
  url: '(https://github.com/furcan/BetterBoard)',
  license: 'MIT Licence (https://opensource.org/licenses/MIT)',
  year: new Date().getFullYear() || '2022',
  cleanCSSOptions,
  minifyOptions,
  minifyOverrides,
  terminalError,
};
// - Exports: end
// Constants: end


index.d.ts
/*!
* BetterBoard - Virtual Keyboard ('https://github.com/Sid220/BetterBoard')
* Version: 1.0.0-beta.01
* Author: The Fake Slim Shady ('https://github.com/Sid220')
* Copyright 2022 BetterBoard - Virtual Keyboard, MIT Licence ('https://opensource.org/licenses/MIT')*
*/

/**
  * BetterBoard is a pure JavaScript library for using virtual keyboards.
  * @namespace BetterBoard
  * @memberof Global
  */
declare namespace BetterBoard {

  /**
   * @interface IBetterBoardOptions
   * @memberof BetterBoard
   */
  export interface IBetterBoardOptions {

    /**
     * @property {Array<Object>} - Optional, An Array of Objects has to be defined for the custom keys.
     * @defaultValue `[ { "0": "Q", "1": "W", "2": "E", "3": "R", "4": "T", "5": "Y", "6": "U", "7": "I", "8": "O", "9": "P" }, { "0": "A", "1": "S", "2": "D", "3": "F", "4": "G", "5": "H", "6": "J", "7": "K", "8": "L" }, { "0": "Z", "1": "X", "2": "C", "3": "V", "4": "B", "5": "N", "6": "M" } ]`
     *
     * e.g. [{"key":"value"}, {"key":"value"}] => [{"0":"A","1":"B","2":"C"}, {"0":"D","1":"E","2":"F"}]
     *
     * Hint: Each object creates a row element (HTML) on the keyboard.
     */
    keysArrayOfObjects?: { [index: string]: string }[];

    /**
    * @property {string} - Required, The default type of keyboard to be shown (can be overridden with data-betterboard-type).
    * @defaultValue `all`
    */
    keyboardType: 'all' | 'keyboard' | 'numpad';

    /**
    * @property {string} - Optional, The position of the keyboard (can be overridden with data-betterboard-placement).
    * @defaultValue `all`
    */
    keyboardPlacement?: 'top' | 'bottom';

    /**
     * @property {string} - Optional
     * @defaultValue `null`
     *
     * The path of the "BetterBoard-keys-${langugage}.json" file must be set to the "keysJsonUrl" option. (XMLHttpRequest to get the keys from JSON file.)
     *
     * e.g. '/Content/Plugins/BetterBoard/dist/BetterBoard-keys-english.json'
     */
    keysJsonUrl?: string;

    /**
     * @property {boolean} - Optional, Whether or not to show the special character button, data-betterboard-specialcharacters can override. Shown when `true`.
     * @defaultValue `true`
     *
     */
    specialCharButton?: boolean;

    /**
    * @property {boolean} - Optional, Whether or not to show the close keyboard button, Shown when `true`.
    * @defaultValue `true`
    *
    */
    closeKeyboardButton?: boolean;

    /**
     * @property {Array<string>} - Optional, An Array of Strings can be set to override the built-in special characters.
     * @defaultValue `null`
     *
     * e.g. ["#", "€", "%", "+", "-", "*"]
     */
    keysSpecialCharsArrayOfStrings?: string[];

    /**
     * @property {Array<number>} - Optional, An Array of Numbers can be set to override the built-in numpad keys. (From 0 to 9, in any order.)
     * @defaultValue `null`
     *
     * e.g. [1, 2, 3, 4, 5, 6, 7, 8, 9, 0]
     */
    keysNumpadArrayOfNumbers?: number[];

    /**
     * @property {string} - Optional, Language Code `(ISO 639-1)` for custom keys (for language support).
     * @defaultValue `en`
     *
     * e.g. `de`, `en`, `fr`, `hu`, `tr`, etc...
     */
    language?: string;

    /**
     * @property {string} - Optional, The theme of keyboard.
     * @defaultValue `light`
     */
    theme?: 'light' | 'dark' | 'flat' | 'material' | 'oldschool';

    /**
     * @property {boolean} - Optional, Uppercase or lowercase to start. Uppercased when `true`.
     * @defaultValue `false`
     */
    capsLockActive?: boolean;

    /**
     * @property {boolean} - Optional, Allow or prevent real/physical keyboard usage. Prevented when `false`.
     * @defaultValue `true`
     *
     * In addition, the `allowMobileKeyboard` option must be `true` as well, if the real/physical keyboard has wanted to be used.
     */
    allowRealKeyboard?: boolean;

    /**
     * @property {boolean} - Optional, Allow or prevent mobile keyboard usage. Prevented when `false`.
     * @defaultValue `true`
     */
    allowMobileKeyboard?: boolean;

    /**
     * @property {boolean} - Optional, CSS animations for opening or closing the keyboard.
     * @defaultValue `true`
     */
    cssAnimations?: boolean;

    /**
     * @property {number} - Optional, CSS animations duration as millisecond.
     * @defaultValue `360`
     */
    cssAnimationsDuration?: number;

    /**
     * @property {string} - Optional, CSS animations style for opening or closing the keyboard.
     * @defaultValue `slide`
     */
    cssAnimationsStyle?: 'slide' | 'fade';

    /**
     * @property {boolean} - Optional, Enable or Disable Spacebar functionality on the keyboard. The Spacebar will be passive when `false`.
     * @defaultValue `true`
     */
    keysAllowSpacebar?: boolean;

    /**
     * @property {string} - Optional, Text of the space key (Spacebar). Without text => `" "`
     * @defaultValue `Space`
     */
    keysSpacebarText?: string;

    /**
     * @property {string} - Optional, Font family of the keys.
     * @defaultValue `sans-serif`
     */
    keysFontFamily?: string;

    /**
     * @property {string} - Optional, Font size of the keys.
     * @defaultValue `22px`
     */
    keysFontSize?: string;

    /**
     * @property {string} - Optional, Font weight of the keys.
     * @defaultValue `normal`
     */
    keysFontWeight?: string;

    /**
     * @property {string} - Optional, Size of the icon keys.
     * @defaultValue `25px`
     */
    keysIconSize?: string;

    /**
     * @property {boolean} - Optional, Scrolls the document to the top or bottom(by the placement option) of the input/textarea element. Prevented when `false`.
     * @defaultValue `true`
     */
    autoScroll?: boolean;
  }

  /**
    * This method can be used to set custom options globally for BetterBoard.
    * @function init
    * @memberof BetterBoard
    * @param {Object} initOptions - Required, `BetterBoard.IBetterBoardOptions`.
    */
  function init(initOptions: IBetterBoardOptions): void;

  /**
    * This method can be used to use BetterBoard on the selected `input` and/or `textarea` elements.
    * @function run
    * @memberof BetterBoard
    * @param {string | HTMLInputElement | HTMLTextAreaElement } selectorOrElement - Required, CSS selector(s) that matches the element(s) or an `input`/`textarea` element.
    * @param {Object} initOptions - Required, `BetterBoard.IBetterBoardOptions`.
    */
  function run(
    selectorOrElement: string | HTMLInputElement | HTMLTextAreaElement,
    initOptions: IBetterBoardOptions,
  ): void;

}

export = BetterBoard;
export as namespace BetterBoard;
